### ISSUE TITLE
[maybe a numpy bug] subtracting two np.ndarrays is of type NoReturn

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#2696]
pyright version: 1.1.195
numpy version: 1.21.4

```py
import numpy as np

reveal_type(np.zeros(1) - np.zeros(1))
```
pyright: Type of "np.zeros(1) - np.zeros(1)" is "NoReturn"
mypy: Revealed type is "Any"

**Comment by erictraut** [comment]
The problem is that `np.zeros` is annotated to return the type `np.ndarray`. But `ndarray` is a generic class that takes two type arguments. By omitting these, its type is implicitly `np.ndarray[Any, Any]`. This type matches the `self` and `other` parameter in the following `__sub__` method overload:

```python
    @overload
    def __sub__(self: NDArray[bool_], other: _ArrayLikeBool_co) -> NoReturn: ...
```

This overload returns `NoReturn`, which is why you're seeing the results that you're seeing.

I'm not sure why mypy is not providing the same results. When I try to run the latest published verison of mypy (0.910) on your sample above it hangs indefinitely. It doesn't seem to like `import numpy as np`. I'm not sure what's going on here because mypy is working fine for me on other files.

As a workaround, you can create a temporary variable with an annotated type for the result of `np.zeros`. Something like this:
```python
import numpy as np
import numpy.typing as npt

x1: npt.NDArray[np.float64] = np.zeros(1)
x2: npt.NDArray[np.float64] = np.zeros(1)

reveal_type(x1 - x2)
```

You may want to report this issue to numpy repo to see if there are improvements they can make to their type annotations to fix this issue. The maintainers of numpy have been making a big investment in types over the past year.

**Comment by twoertwein** [comment]
Thank you, for your analysis! I opened an issue at numpy.
